File: questik/app/src/main/java/ru/rinet/questik/data/room/model/CategoryDao.java
package ru.rinet.questik.data.room.model;

import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.Query;
import androidx.room.Transaction;
import androidx.room.Update;

import java.util.List;

@Dao
public interface CategoryDao {

    @Query("SELECT * FROM category")
    List<Category> getAll();

    @Query("SELECT * FROM category WHERE _id = :id")
    Category getById(int id);
    @Transaction
    @Query("SELECT * FROM category WHERE _id IN (:categoryWithJob)")
    List<CategoryWithJob> getCategoryWithJob(int categoryWithJob);

    @Insert
    void insert(Category category);


    @Update
    void update(Category category);

    @Delete
    void delete(Category category);
}

File: app/src/main/java/ru/rinet/questik/repo/local/room/dao/CategoryDao.kt
package ru.rinet.questik.repo.local.room.dao

import androidx.paging.PagingSource
import androidx.room.*
import ru.rinet.questik.repo.local.room.entity.CategoryEntity

@Dao
interface CategoryDao {

    @Query("SELECT COUNT(id) FROM categories")
    fun getCategoriesCount(): Int

    @Query("SELECT * FROM categories")
    fun getAllCategories(): PagingSource<Int, CategoryEntity>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insertAllCategories(items:List<CategoryEntity>)

    @Query("DELETE FROM categories")
    fun clearCategories()

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insertCategory(category: CategoryEntity)

    @Update
    fun updateCategory(category: CategoryEntity)

    @Delete
    fun deleteCategory(category: CategoryEntity)

    @Query("SELECT * FROM categories WHERE id == :id")
    fun getCategoryById(id: Int): List<CategoryEntity>

    @Query("SELECT * FROM categories")
    fun getCategories(): List<CategoryEntity>
}